aten/src/ATen/native/cuda/AtomicOps.cuh

```diff
@@ -0,0 +1,344 @@
+
+#pragma once
+
+#include <c10/util/complex.h>
+#include <THC/THC.h>
+#include <TH/THHalf.h>
+#include <THC/THCNumerics.cuh>
+#include <ATen/ATen.h>
+
+namespace at { namespace native { namespace atomic_ops {
+
+template <typename operation, size_t n>
+struct AtomicIntegerop;
+
+using add_op = std::integral_constant<int, 0>;
+using mul_op = std::integral_constant<int, 1>;
+
+// Integer addition functions.
+template<typename T>
+struct AtomicIntegerop<T, 1> {
+  template <typename func_t>
+  inline __device__ void operator()(T *address, T val, const func_t& func) {
+    size_t offset = (size_t)address & 3;
+    uint32_t * address_as_ui = (uint32_t *)((char *)address - offset);
+    uint32_t old = *address_as_ui;
+    uint32_t shift = offset * 8;
+    uint32_t old_byte;
+    uint32_t newval;
+    uint32_t assumed;
+
+    do {
+      assumed = old;
+      old_byte = (old >> shift) & 0xff;
+      // preserve size in initial cast. Casting directly to uint32_t pads
+      // negative signed values with 1's (e.g. signed -1 = unsigned ~0).
+      newval = func(val, old_byte);//static_cast<uint8_t>(THCNumerics<T>::add(val, old_byte));
+      newval = (old & ~(0x000000ff << shift)) | (newval << shift);
+      old = atomicCAS(address_as_ui, assumed, newval);
+    } while (assumed != old);
+  }
+};
+
+template<typename T>
+struct AtomicIntegerop<T, 2> {
+  template <typename func_t>
+  inline __device__ void operator()(T *address, T val, const func_t& func) {
+    size_t offset = (size_t)address & 2;
+    uint32_t * address_as_ui = (uint32_t *)((char *)address - offset);
+    bool is_32_align = offset;
+    uint32_t old = *address_as_ui;
+    uint32_t old_bytes;
+    uint32_t newval;
+    uint32_t assumed;
+
+    do {
+      assumed = old;
+      old_bytes = is_32_align ? old >> 16 : old & 0xffff;
+      // preserve size in initial cast. Casting directly to uint32_t pads
+      // negative signed values with 1's (e.g. signed -1 = unsigned ~0).
+      newval = func(val, old_bytes);//static_cast<uint16_t>(THCNumerics<T>::add(val, old_bytes));
+      newval = is_32_align ? (old & 0xffff) | (newval << 16) : (old & 0xffff0000) | newval;
+      old = atomicCAS(address_as_ui, assumed, newval);
+    } while (assumed != old);
```

Comment:
Could you please provide a reference for these implementations. If I understand well, these implementation extends `atomicAdd` function for types that are not covered by specific CUDA architecture/version.

Reply:
They're derived from THCAtomics which was already present in the pytorch source code. In time, I think these routines can replace the ones in THCAtomics since these are more generic. 

https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCAtomics.cuh